src/freenet/support/MultiValueTable.java

```diff
@@ -1,230 +1,260 @@
  */
-
 public class MultiValueTable<K,V> {
-    private Hashtable<K, Vector<V>> table;
-    private int ies;
+    private final Map<K, List<V>> table;
 
     public MultiValueTable() {
-        this(16, 3);
+        table = new ConcurrentHashMap<>();
     }
 
     public MultiValueTable(int initialSize) {
-        this(initialSize, 3);
+        table = new ConcurrentHashMap<>(initialSize);
     }
 
-    public MultiValueTable(int initialSize, int initialEntrySize) {
-        table = new Hashtable<K, Vector<V>>(initialSize);
-        ies = initialEntrySize;
+    /**
+     * Deprecated constructor, please use other variant with only one single size parameter
+     * @param initialSize table initial size
+     * @param unused this parameter was used as initial value collection size, and it is not unused anymore.
+     *               Values collections grow according to Java collection rules.
+     */
+    @Deprecated
+    public MultiValueTable(int initialSize, int unused) {
+        this(initialSize);
     }
 
     public static <K, V> MultiValueTable<K, V> from (K[] keys, V[] values) {
-      if (keys.length != values.length) {
-        throw new IllegalArgumentException(String.format(
-            "keys and values must contain the same number of values, but there are %d keys and %d values",
-            keys.length,
-            values.length));
-      }
-      MultiValueTable<K, V> table = new MultiValueTable<>();
-      for (int i = 0; i < keys.length; i++) {
-        table.put(keys[i], values[i]);
-      }
-      return table;
+        if (keys.length != values.length) {
+            throw new IllegalArgumentException(String.format(
+                "keys and values must contain the same number of values, but there are %d keys and %d values",
+                keys.length,
+                values.length));
+        }
+        MultiValueTable<K, V> table = new MultiValueTable<>(keys.length);
+        for (int i = 0; i < keys.length; i++) {
+            table.put(keys[i], values[i]);
+        }
+        return table;
+    }
+
+    @SafeVarargs
+    public static <K, V> MultiValueTable<K, V> from(K key, V... values) {
+        return from(key, Arrays.asList(values));
+    }
+
+    public static <K, V> MultiValueTable<K, V> from(K key, Collection<V> values) {
+        MultiValueTable<K, V> table = new MultiValueTable<>(1);
+        table.putAll(key, values);
+        return table;
     }
 
     public void put(K key, V value) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            if (v == null) {
-                v = new Vector<V>(ies);
-                table.put(key, v);
+        this.table.compute(key, (k, list) -> {
+            if (list == null) {
+                list = new CopyOnWriteArrayList<>();
             }
-            v.addElement(value);
-        }
+            list.add(value);
+            return list;
+        });
+    }
+
+    public void putAll(K key, Collection<V> elements) {
```

Comment:
`Collection<V>` put arguments should read `Collection<? extends V>` instead, so it becomes possible to put a `List<Foo>` into a `MultiValueTable<?, Object>`.

Reply:
Pushed changes for `putAll()` and `from()` methods.